adafruit_ble/current_time_client.py

```diff
@@ -0,0 +1,147 @@
+ # The MIT License (MIT)
+#
+# Copyright (c) 2019 Dan Halbert for Adafruit Industries
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_ble.current_time_client`
+====================================================
+
+UART-style communication by a Central as a GATT Client
+
+* Author(s): Dan Halbert for Adafruit Industries
+
+"""
+import struct
+import time
+
+from bleio import Peripheral, UUID
+from .advertising import SolicitationAdvertisement
+
+class CurrentTimeClient:
+    """
+    Set up a peripheral that solicits centrals for Current Time Service.
+
+    :param str name: Name to advertise for server. If None, use default Advertisement name.
+
+    Example::
+
+        from adafruit_ble.current_time_client import CurrentTimeClient
+
+        cts_client = CurrentTimeClient()
+        cts_client.start_advertising()
+        while not cts_client.connected:
+            pass
+        cts_client.discover()
+        cts_client.pair()
```

Comment:
Is pairing required? Could you discover and then read the property?

It's not clear to me why discover is separate from pair or the property read. Seems like it could be done internally as needed instead.

Reply:
Pairing is required for this particular service. I left these as explicit because the client may want to do stuff at certain times (e.g., may do other stuff while waiting for a connection). Discovery can take several seconds.

I could have an internal flag `self._discovery_done` or similar and do discovery (and then pairing) when one of the value properties is read. Then the first read will take several seconds. I left it flexible for now.

Reply:
Thinking more, on the other hand, the user program can always get the time right away, and expect it to take a little while. OK, I'll hide these.